src/FFmpegWriter.cpp

```diff
@@ -910,60 +914,63 @@ void FFmpegWriter::flush_encoders() {
-			write_audio_count += av_rescale_q(audio_input_position / (audio_codec_ctx->channels * av_get_bytes_per_sample(AV_SAMPLE_FMT_S16)), av_make_q(1, info.sample_rate), audio_codec_ctx->time_base);
-
-			AVPacket pkt;
-			av_init_packet(&pkt);
-			pkt.data = NULL;
-			pkt.size = 0;
-			pkt.pts = pkt.dts = write_audio_count;
+	if (info.has_audio) {
+        for (;;) {
+            AVPacket pkt;
+            av_init_packet(&pkt);
+            pkt.data = NULL;
+            pkt.size = 0;
+            pkt.pts = pkt.dts = write_audio_count;
 
-			/* encode the image */
-			int error_code = 0;
-			int got_packet = 0;
+            /* encode the image */
+            int error_code = 0;
+            int got_packet = 0;
 #if IS_FFMPEG_3_2
-			error_code = avcodec_send_frame(audio_codec_ctx, NULL);
+            error_code = avcodec_send_frame(audio_codec_ctx, NULL);
 #else
-			error_code = avcodec_encode_audio2(audio_codec_ctx, &pkt, NULL, &got_packet);
+            error_code = avcodec_encode_audio2(audio_codec_ctx, &pkt, NULL, &got_packet);
 #endif
-			if (error_code < 0) {
-				ZmqLogger::Instance()->AppendDebugMethod("FFmpegWriter::flush_encoders ERROR [" + (std::string)av_err2str(error_code) + "]", "error_code", error_code);
-			}
-			if (!got_packet) {
-				break;
-			}
+            if (error_code < 0) {
+                ZmqLogger::Instance()->AppendDebugMethod(
+                        "FFmpegWriter::flush_encoders ERROR [" + (std::string) av_err2str(error_code) + "]",
+                        "error_code", error_code);
+            }
+            if (!got_packet) {
+                break;
+            }
 
-			// Since the PTS can change during encoding, set the value again.  This seems like a huge hack,
-			// but it fixes lots of PTS related issues when I do this.
-			pkt.pts = pkt.dts = write_audio_count;
+            // Since the PTS can change during encoding, set the value again.  This seems like a huge hack,
+            // but it fixes lots of PTS related issues when I do this.
+            pkt.pts = pkt.dts = write_audio_count;
 
-			// Scale the PTS to the audio stream timebase (which is sometimes different than the codec's timebase)
-			if (pkt.pts != AV_NOPTS_VALUE)
-				pkt.pts = av_rescale_q(pkt.pts, audio_codec_ctx->time_base, audio_st->time_base);
-			if (pkt.dts != AV_NOPTS_VALUE)
-				pkt.dts = av_rescale_q(pkt.dts, audio_codec_ctx->time_base, audio_st->time_base);
-			if (pkt.duration > 0)
-				pkt.duration = av_rescale_q(pkt.duration, audio_codec_ctx->time_base, audio_st->time_base);
+            // Scale the PTS to the audio stream timebase (which is sometimes different than the codec's timebase)
+            if (pkt.pts != AV_NOPTS_VALUE)
+                pkt.pts = av_rescale_q(pkt.pts, audio_codec_ctx->time_base, audio_st->time_base);
+            if (pkt.dts != AV_NOPTS_VALUE)
+                pkt.dts = av_rescale_q(pkt.dts, audio_codec_ctx->time_base, audio_st->time_base);
```

Comment:
Shouldn't these be using [`av_packet_rescale_ts()`](https://ffmpeg.org/doxygen/4.1/group__lavc__packet.html#gae5c86e4d93f6e7aa62ef2c60763ea67e) instead?

Reply:
Actually, the really nice thing about `av_packet_rescale_ts()` is that it does _all_ the timestamp values at once, and auto-ignores `AV_NOPTS_VALUE`.

So, instead of all this:
```c++
            if (pkt.pts != AV_NOPTS_VALUE)
                pkt.pts = av_rescale_q(pkt.pts, audio_codec_ctx->time_base, audio_st->time_base);
            if (pkt.dts != AV_NOPTS_VALUE)
                pkt.dts = av_rescale_q(pkt.dts, audio_codec_ctx->time_base, audio_st->time_base);
            if (pkt.duration > 0)
                pkt.duration = av_rescale_q(pkt.duration, audio_codec_ctx->time_base, audio_st->time_base);
```
just:
```c++
           av_packet_rescale_ts(&pkt, audio_codec_ctx->time_base, audio_st->time_base);
```

A [relevant mailing list thread](http://ffmpeg.org/pipermail/libav-user/2018-January/010843.html).

Reply:
Thanks, this is great!